drivers/gps/nrf9160_gps/nrf9160_gps.c

```diff
@@ -204,173 +237,56 @@ static void notify_event(const struct device *dev, struct gps_event *evt)
-	 * implicitly stops the GPS.
-	 */
-	retval = nrf_setsockopt(drv_data->socket,
-				NRF_SOL_GNSS,
-				NRF_SO_GNSS_START,
-				&delete_mask,
-				sizeof(delete_mask));
-	if (retval != 0) {
-		LOG_ERR("Failed to start GPS");
-		return -EIO;
-	}
-
-	return 0;
+	k_work_cancel_delayable(&drv_data->timeout_work);
+	k_work_cancel_delayable(&drv_data->blocked_work);
 }
 
 static void gps_thread(int dev_ptr)
 {
 	struct device *dev = INT_TO_POINTER(dev_ptr);
 	struct gps_drv_data *drv_data = dev->data;
-	int len;
-	bool operation_blocked = false;
-	bool has_fix = false;
-	struct gps_event evt = {
-		.type = GPS_EVT_SEARCH_STARTED
-	};
-
-wait:
-	k_sem_take(&drv_data->thread_run_sem, K_FOREVER);
-
-	evt.type = GPS_EVT_SEARCH_STARTED;
-	notify_event(dev, &evt);
+	struct event_item event;
+	struct gps_event evt;
+	struct nrf_modem_gnss_pvt_data_frame *pvt_data;
+	struct nrf_modem_gnss_nmea_data_frame *nmea_data;
+	struct nrf_modem_gnss_agps_data_frame *agps_data;
+	size_t nmea_len;
 
 	while (true) {
-		nrf_gnss_data_frame_t raw_gps_data = {0};
-		struct gps_event evt = {0};
-
-		/* There is no way of knowing if nrf_recv() blocks because the
-		 * GPS timeout/retry value has expired or the GPS has gotten a
-		 * fix. This check makes sure that a GPS_EVT_SEARCH_TIMEOUT is
-		 * not propagated upon a fix.
-		 */
-		if (!has_fix) {
-			/* If nrf_recv() blocks for more than one second,
-			 * a fix has been obtained or the GPS has timed out.
-			 * Schedule a delayed timeout work to be executed after
-			 * five seconds to make sure the appropriate event is
-			 * propagated upon a block.
-			 */
-			k_work_schedule(&drv_data->timeout_work, K_SECONDS(5));
-		}
+		k_msgq_get(&event_msgq, &event, K_FOREVER);
 
-		len = nrf_recv(drv_data->socket, &raw_gps_data,
-			       sizeof(nrf_gnss_data_frame_t), 0);
+		switch (event.id) {
+		case NRF_MODEM_GNSS_EVT_PVT:
+			pvt_data = event.data;
 
-		/* At this point, it may be too late to cancel this work,
-		 * as it may already be running. In such case, the search
-		 * timeout notification will be generated anyway.
-		 */
-		k_work_cancel_delayable_sync(&drv_data->timeout_work,
-					     &drv_data->work_sync);
-
-		if (len <= 0) {
-			/* Is the GPS stopped, causing this error? */
-			if (!atomic_get(&drv_data->is_active)) {
-				goto wait;
-			}
-
-			if (errno == EHOSTDOWN) {
-				LOG_DBG("GPS host is going down, sleeping");
-				cancel_works(drv_data, true);
-				atomic_clear(&drv_data->is_active);
-				atomic_set(&drv_data->is_shutdown, 1);
-				nrf_close(drv_data->socket);
-
-				nrf_modem_lib_shutdown_wait();
-				if (open_socket(drv_data) != 0) {
-					LOG_ERR("Failed to open socket after "
-						"shutdown sleep, killing thread");
-					return;
-				}
-
-				atomic_clear(&drv_data->is_shutdown);
-				LOG_DBG("GPS host available, going back to "
-					"initialized state");
-				goto wait;
+			if (pvt_data->flags & NRF_MODEM_GNSS_PVT_FLAG_FIX_VALID) {
+				k_work_cancel_delayable(&drv_data->timeout_work);
```

Comment:
Would be better to call _k_work_cancel_delayable_sync()_ instead as it was done earlier. This is a executed on the context of the GPS thread so no need to worry about stalling an event queue.

Reply:
Can you explain why `k_work_cancel_delayable_sync()` should be used here? The only difference seems to be that the cancellation is performed synchronously, i.e. the calling thread waits until the delayable work item has been cancelled. I don't see the reason for that here.

Reply:
> A large number of problems with Zephyr's work queue result from the decision to support cancelling work items that have been submitted, while failing to account for cases where the cancellation could not be performed immediately. It's years too late to change that, but I strongly suspect that a better design would have justified cancelling a work item only in a situation where a subsystem that uses the work queue is being shut down. In that case, you want to cancel, and confirm cancellation, to be sure it's safe to release any resources in the subsystem---or that it's safe to restart it. This is what synchronous cancellation is for.
> 
> It comes down to a question: If you cancel a work item, do you mean to cancel all past submissions/schedules of that item, or do you want to force the work item to an idle state where its handler will not be invoked again? These are very different goals. For the latter, resubmission of a still-running item must be rejected otherwise it may never become idle. For the former, you could probably accomplish the same thing by recording additional state specific to the work item and the events that cause it to be scheduled/submitted.

https://github.com/zephyrproject-rtos/zephyr/issues/35255#issuecomment-840877843

Reply:
One more thing to consider is that you can't resubmit a work queue item as long as there is one in cancelled state.

````
prevents resubmission of a item being canceled until cancellation completes;
````
https://github.com/zephyrproject-rtos/zephyr/pull/29618#issue-511682580

Reply:
I know there have been problems related to cancelling a work item. I'm just trying to understand the actual issue and how it affects the GPS driver.

My understanding is, that synchronous cancel only blocks when the delayed work item is running (i.e. being executed). In this case it would block until the item stops running. GPS driver uses two delayed works, one sends a timeout event and the second sends one RPC message. Thus, it's unlikely that the execution takes a long time. Work items are (re)scheduled only when a new PVT notification is received from GNSS, i.e. this can happen only once a second. For the synchronous cancel to make any difference, the work being cancelled would have to run for more than one second. Otherwise the cancellation would be completed before the next PVT notification is received anyway.

Also, it seems that scheduling a work item being cancelled is not a problem, submitting it into a work queue is. The code never submits the works directly, but schedules them for execution 5 or 10 seconds later.

https://docs.zephyrproject.org/latest/reference/kernel/threads/workqueue.html#c.k_work_cancel_delayable

_Canceling delayable work does not prevent rescheduling it. It does prevent submitting it until the cancellation completes._

Finally, I don't want the GPS driver thread to be blocked for one second or more. If the thread would be blocked for this long time, the event queue is likely to become full. This would result in PVTs or NMEAs not being processed at all, which would actually be worse than those delayed works not being run as expected.

Reply:
Ok, sounds good.